### PR TITLE
Feat/292 remove tensorflow

### PR DESCRIPTION
**Pull request by em-t**
Remove tensorflow, replace with pytorch.

Sorry this went until the very last minute before I start my summer vacation. It's still incomplete, but should be in pretty good shape for @jtlait or @lehtonenp to finish. The training functions and the prediction functions are implemented. I've ran through them all with some data, but have not gone through all options or tested a wide variety of use cases. Testing functions are missing but the UI is in place for those. There's a lot of things to finish up and polish in the training and prediction functions too, but I've tried to mark missing things with TODO comments so that they would be easy to spot. The main ones include:
- Handle raster label data (validations on the tool UI side for what's allowed is done, but handling on the function-side isn't)
- Multi-class vs. binary classification prediction function should be double-checked
- Early stopping has not been implemented
- SMOTE is not plugged in
- Validation data is not handled

I added quite a lot of validation to the tool UI side (instead of the function internals), since that is much nicer for the user. The UI tries to handle label files in particular better. I also made a couple of quality-of-life improvements that were not in the specs (or were dropped at some point), namely:
- Better control when adding model layers, ability to add activation function and dropout rate per-layer
- Re-added the final layer back to the classifier training tool, because multi-class is now supported (sparse categorical cross-entropy, not one-hot encoded). (More validation should be done on the function-side, however, to make sure combinations of number of output classes, output layer activation and loss function make sense.)
- Allow string labels (i.e. feature data with a string attribute). These are encoded to integers, because many of the ML python functions would not be able to deal with them otherwise, but the mapping is both logged to the user and saved as .json to the scratch dir in case the user has need for it.
- Developer quality-of-life improvement, which adds some extra lines of code but truly makes editing the tool code a bit less annoying: save the tool parameter indices to the tool class so that they don't need to be edited in tons of places, which is prone to mistakes. Also extracted common tool parameter creation into function to reduce places where mistakes can occur due to copying.

Also, added the loss plot requested by @jtlait ! (It's quite plain and I'm not 100% sure it's what you asked for, so you might want to improve it.)

Some final things to be aware of:
- In addition to the model file, the training tools also save another file with metadata. (These could have been combined, but I thought it best to save the model state dict as is, since that's best practice. The metadata is needed, however, when loading the model.)
- Some tool/field/etc. titles and such may not be exactly as specified (mostly relevant for the user manual). I meant to go through them in the end, but ran out of time. Also, any metadata that the tools might have had is gone, since I deleted the old tools before replacing them.
- I did not have time to do much clean-up, so the non-UI code is very rough around the edges and quite repetitive between classification and regression functions. If someone has the time, it would be great to give it a bit of a refactor.

### CODE REVIEW

File: Toolbox/arcsdm/machine_learning/mlp/classification/training.py
Comment (jtlait) [inline Toolbox/arcsdm/machine_learning/mlp/classification/training.py]: We should use val_loss_dict here instead of train_loss_dict

Comment (jtlait) [inline Toolbox/arcsdm/machine_learning/mlp/classification/training.py]: Printout could spescify that validation loss is used "Epoch with smallest validation loss..."

Comment (jtlait) [inline Toolbox/arcsdm/machine_learning/mlp/classification/training.py]: We could add stratify=y into both train_test_split calls to ensure more balanced train_validation splits

Comment (jtlait) [inline Toolbox/arcsdm/machine_learning/mlp/classification/training.py:402]: best model state doesn't often correspond to trained epochs. We could add one extra parameter, best_epoch that describes the epoch when best validation loss was reached

**Review by jtlait — CHANGES_REQUESTED** [review]
Well done! I tested this on ArcGIS Pro and although testing was not extensive, everything seemed to work as expected. I left some minor requests for modifications

**Review by jtlait — APPROVED** [review]
Looks good to me. Nice work!
